File: web/src/types/floorplan.ts
export interface FloorPoint {
  x: number
  y: number
}

export interface FloorWall {
  id: string
  points: FloorPoint[]
  thickness?: number
}

export interface FloorRoom {
  id: string
  name: string
  polygon: FloorPoint[]
  area?: number | null
}

export interface FloorOpening {
  id: string
  type: 'door' | 'window'
  wall_id: string
  position: number
  width: number
  connects: string[]
}

export interface FloorPlanData {
  scale: number | null
  status: string
  walls: FloorWall[]
  rooms: FloorRoom[]
  openings: FloorOpening[]
  source_url?: string | null
  source_width?: number | null
  source_height?: number | null
}

export function polygonCentroid(polygon: FloorPoint[]): FloorPoint {
  if (!polygon.length) {
    return { x: 0, y: 0 }
  }
  const sum = polygon.reduce(
    (acc, point) => ({ x: acc.x + point.x, y: acc.y + point.y }),
    { x: 0, y: 0 },
  )
  return { x: sum.x / polygon.length, y: sum.y / polygon.length }
}

export function computeViewBox(floorplan: FloorPlanData, padding = 20): string {
  if (floorplan.source_width && floorplan.source_height) {
    return `${-padding} ${-padding} ${floorplan.source_width + padding * 2} ${floorplan.source_height + padding * 2}`
  }
  const points: FloorPoint[] = []
  floorplan.walls.forEach((wall) => points.push(...wall.points))
  floorplan.rooms.forEach((room) => points.push(...room.polygon))
  if (!points.length) {
    return '0 0 400 300'
  }
  const xs = points.map((p) => p.x)
  const ys = points.map((p) => p.y)
  const minX = Math.min(...xs) - padding
  const minY = Math.min(...ys) - padding
  const maxX = Math.max(...xs) + padding
  const maxY = Math.max(...ys) + padding
  return `${minX} ${minY} ${maxX - minX} ${maxY - minY}`
}

export function scaleLabel(scale: number | null): string {
  if (!scale) {
    return '未标定'
  }
  const metersPerPixel = 1 / scale
  return `1px = ${metersPerPixel.toFixed(3)}m`
}

File: server/app/schemas/floorplan.py
from enum import Enum
from typing import Annotated, Literal

from pydantic import BaseModel, Field, model_validator

from app.services.floorplan.plan_classifier import PlanType


class FloorPlanStatus(str, Enum):
    DRAFT = "draft"
    CONFIRMED = "confirmed"


class OpeningType(str, Enum):
    DOOR = "door"
    WINDOW = "window"


class ValidationIssue(BaseModel):
    code: str
    severity: Literal["error", "warning", "info"]
    message: str
    room_ids: list[str] = Field(default_factory=list)


class FloorPlanValidation(BaseModel):
    level: Literal["pass", "warning", "error", "unknown"] = "unknown"
    checked_at: str | None = None
    issues: list[ValidationIssue] = Field(default_factory=list)


class ParseMeta(BaseModel):
    vlm_model: str | None = None
    vlm_steps: int | None = None
    cv_wall_quality: float | None = None
    wall_source: Literal["cv", "polygon"] | None = None


class Point(BaseModel):
    x: float
    y: float


class Wall(BaseModel):
    id: str
    points: Annotated[list[Point], Field(min_length=2)]
    thickness: float = 0.2


class Room(BaseModel):
    id: str
    name: str
    polygon: Annotated[list[Point], Field(min_length=3)]
    area: float | None = None


class Opening(BaseModel):
    id: str
    type: OpeningType
    wall_id: str
    position: Annotated[float, Field(ge=0, le=1)]
    width: float = Field(gt=0)
    connects: list[str] = Field(default_factory=list)


class FloorPlanModel(BaseModel):
    scale: float | None = Field(default=None, gt=0)
    walls: list[Wall] = Field(default_factory=list)
    rooms: list[Room] = Field(default_factory=list)
    openings: list[Opening] = Field(default_factory=list)
    status: FloorPlanStatus = FloorPlanStatus.DRAFT
    plan_type: PlanType | None = None
    validation: FloorPlanValidation | None = None
    parse_meta: ParseMeta | None = None

    @model_validator(mode="after")
    def validate_walls_for_confirmed(self) -> "FloorPlanModel":
        if self.status == FloorPlanStatus.CONFIRMED and not self.walls:
            raise ValueError("walls must not be empty when status is confirmed")
        return self


class FloorPlanRead(FloorPlanModel):
    source_image: str | None = None
    original_filename: str | None = None
    estimated_area: float | None = None
    source_url: str | None = None
    source_width: int | None = None
    source_height: int | None = None
    has_watermark: bool | None = None
    plan_type_label: str | None = None
    plan_type_message: str | None = None


class ScaleRequest(BaseModel):
    point_a: Point
    point_b: Point
    distance_m: float = Field(gt=0)
